### PR TITLE
Ensure to escape characters before constructing JSON profile trace

### LINKED ISSUE #21858: `-Yprofile-trace` doesn't correctly escape output
## Compiler version

3.6.2-RC1-bin-20241028-a3786a5-NIGHTLY

## Minimized code

Compiling this code with `-Yprofile-enabled -Yprofile-trace test.trace`

```Scala
class /\
```

## Output

Results in a `test.trace` file that contains invalid JSON since the backslash in `class /\` is not escaped:

```json
{"cat":"typecheck","name":"class /\","ph":"E","pid":"22181-00129","tid":"00129","ts":2132512955562}
```

## Expectation

The backslash should be escaped with another backslash, i.e.

```json
{"cat":"typecheck","name":"class /\\","ph":"E","pid":"22181-00129","tid":"00129","ts":2132512955562}
```

### PR DESCRIPTION
**Pull request by WojciechMazur**
Fixes #21858 by setting up special escapes for characters that might  corrupt the output JSON file produced by `-Yprofile-trace`

### CODE REVIEW

**Review by prolativ — APPROVED** [review]
Otherwise LGTM

**Review by prolativ — APPROVED** [review]

File: compiler/src/dotty/tools/dotc/profile/JsonNameTransformer.scala [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    var i = 0
+    while (i < len) {
+      val c = name(i)
+      if (c < nops && (op2code(c.toInt) ne null)) {
+        if (buf eq null) {
+          buf = new StringBuilder()
+          buf.append(name.subSequence(0, i))
+        }
+        buf.append(op2code(c.toInt))
+      } else if (c <= 0x1F || c > 0x7F) {
```
Comment (mbovel) [inline compiler/src/dotty/tools/dotc/profile/JsonNameTransformer.scala:33]: Just to check: is it indeed okay not to escape the DEL character?

I looked at [uJSON's `escapeByte`](https://github.com/com-lihaoyi/upickle/blob/fe53cf3b018aea5df232a71d4a49332781a91dab/upickle/core/src/upickle/core/RenderUtils.scala#L29) to try to get an answer, but it's not clear to me why DEL is excluded when `escapeUnicode` is true but included when it's false.

**Review by mbovel — APPROVED** [review]
It seems a bit crazy to me that we need to rewrite a JSON escape function in 2024, but I guess this goes with the "no-dependency" policy of Dotty, and the fact that neither the Java nor the Scala standard libraries contain such function.

Otherwise looks okay to me!

File: compiler/src/dotty/tools/dotc/profile/JsonNameTransformer.scala (hunk 1 above)
Comment (mbovel) [reply compiler/src/dotty/tools/dotc/profile/JsonNameTransformer.scala:33]: I checked: the DEL character (U+007F) is non-printable, and both Node.js and Python escape it, even though they don’t escape all non-ASCII characters.

```
➜  ~ node                  
Welcome to Node.js v22.8.0.
Type ".help" for more information.
> JSON.stringify("\u007F")
'"\x7F"'
➜  ~ python3
Python 3.12.5 (main, Aug  6 2024, 19:08:49) [Clang 15.0.0 (clang-1500.3.9.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import json
>>> json.dumps("\u007F")
'"\\u007f"'
```

Comment (mbovel) [reply compiler/src/dotty/tools/dotc/profile/JsonNameTransformer.scala:33]:
```suggestion
      } else if (c <= 0x1F || c >= 0x7F) {
```

**Pull request merged**